config/defaults.ini.php

```diff
@@ -0,0 +1,487 @@
+<?php return <<<INI
+
+; This file declares the default values for all the config values of Friendica.
+; Please don't edit this file as its content may change in the upcoming versions.
```

Comment:
This files ensures that config keys exists with a meaningful default value without having to specify it in the code. My goal was to make the `doc/htconfig.md` file obsolete as we would have all the info about config key and default values in the same place.

Reply:
So shouldn't we then remove most content from that documentation file? I'm a lazy coder. I don't want to document at two places when a single place is enough.

Reply:
Me neither, do you want me to do it right now? There's an even more painstaking task to remove all default values from `Config::get()` calls.

Reply:
If @tobiasd has got no objections, please do so.

Concerning the default values: Does this file contain **every** config value, even the ones that are stored in the database and are not changeable via config file?

Reply:
The goal is for this file to contain every config key that is accessed in the code, except for generated ones (like dbupdate_xxx).

Reply:
Then how does an admin know the difference between a config that has got a frontend interface and a config value that can only be changed in the config file?

Reply:
TBH the htconfig or config only variables should die. They should all be accessible from the admin panel. I'm also expecting that over time the "central" file for default values will miss some updates as the default values will be set in the code elsewhere by some lazy coders.

Reply:
Well, for starters the database config (credentials + specific database logging config) will never ever been looked up in the `config` table for cyclic dependency reasons. So there always will be file-only config keys.

Secondly, the `defaults.ini.php` isn't strictly needed, we'll always check for the `App->config` key existence anyway. It just is a meaningful piece of documentation because it is "live". If the default value for a config key only is defined in this file, then we have to change this file to change it, and it will de facto change the documentation as well, no need to update in multiple places.

At least that's the end goal.

Reply:
That is not what I meant with *the file has to die*. Looking at the htconfig documentation file, 99% of the config vars mentioned there should be accessible in the admin panel. For those I think it has to die. Then the documentation file for it would be unnecessary as well, as the documentation would be in the admin panel documentation.

Reply:
I don't feel comfortable with this idea, I think the config key documentation should be available without having to have access to the admin panel. I agree that most of the config should be editable from the admin interface but that there should be a non-HTTP, non-database way of doing it as well. Hence the existence of `defaults.ini.php` .

Reply:
Totally agreed. The *admin panel documentation* I referred to was the `/doc/Settings.md` file. Sorry for being unclear, I did not mean the help text beneath the options in the admin panel.

Additionally there should be documentation available in easy language. For those in need in their native tongue, and those who are happy to manage some broken English.

Reply:
Question is how the config class now behaves. Currently the settings in the ".htconfig.php" only work when they never had been set via the web frontend, since the config class always looks into the database first.

Reply:
The priority order for a config value is:
1. `config/defaults.ini.php` file
2. `config/local.ini.php` file
3. addon config via `load_config` hook
4. `config/addon.ini.php` file
5. `config` table

Each level in order can and will overwrite values that have been set in the previous levels.

Reply:
How fast is the loading of this large ini file - and how much memory does it use?

Reply:
According to phpFiddle Lite:
- 0.0001 s
- 16 712 bytes

Test code:
```php
$start_mem = memory_get_usage();
$ini_string = <<<INI

[Content of defaults.ini.php returned string]

INI;

$start = microtime(true);
$config = parse_ini_string($ini_string);
var_dump(microtime(true) - $start);
var_dump(memory_get_usage() - $start_mem);
```

Reply:
@tobiasd the htconfig only settings are for stuff that is still under testing or not for general use. So I think it is important to have settings without a frontend. And so I have an important question for @MrPetovan: How do we separate the settings that can only be set via config file and the ones that can't be set via config file, but via web frontend?

Reply:
> How do we separate the settings that can only be set via config file and the ones that can't be set via config file, but via web frontend?

Any config value can be set by config file. A limited set can be overridden from the web frontend. To answer your question, if you see it in the web frontend, you can change it.

Reply:
Sorry, but I hope the *still under testing* is  valid only for the options listed under *experimental* in the documentation otherwise this section is superfluous and the content should not be separated.

Reply:
There's no way to separate both, because we have been indiscriminately using either `system` and `config` as categories, sometimes confusingly so. If we need to separate both, I'd suggest using "config" for config file only keys, and "system" or something more meaningful for the ones that can be changed from the web frontend.

Reply:
Then you have your answer.

Reply:
What about not only having the ```defaults.ini.php``` but a second one that is dedicated to the "config file only" settings? That should make things clearer. Thing is, that we need to be able to move settings from the "beta section" to the official section.

Reply:
It is possible, and we would just need to move the default value from a file to another when we create the form field in the admin settings.

I suggest `config/config.ini.php` for the file-only config keys while `config/settings.ini.php` will keep the default value for the admin-controlled settings.

Is the current list of file-only config keys in `doc/htconfig.md` exhaustive?

Reply:
This list is complete. I always keep track on removing the entries there upon moving them to the web settings.

@tobiasd do you agree with this solution?

Reply:
I went ahead with the idea, it was very easy to do so because the config is handled at a single place now. 😇

Reply:
If all but the currently 1 experimental setting eventually arrive in the admin panel.

Reply:
Moving configuration keys around should be done in another PR though.

Reply:
@tobiasd I think that most of the settings in that file are not for general purpose. And before moving them all to the admin panel, we definitely have to rework it. By now it is already much too huge.

Reply:
Following that logic, two thirds of the site configuration can be removed from the admin panel I think. Most of what currently is htconfig only is equivalent to the majority of obscure settings that are already included.

Reply:
But can we wait with moving more stuff to the site config until this page is restructured?

Reply:
Why wait for after? Why not include those *when* restructuring the site configuration page?

Reply:
I would like to delegate that stuff. There are so much tasks for the backend that I personally prefer.Hopefully next month my item changes will be at some final state for this version. After this I would like to focus on AP.